### PR TITLE
Add a generic ItemNormalizer (relation support for raw JSON, XML, CSV, YAML)

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

File: src/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public function denormalize($data, $class, $format = null, array $context = [])
+    {
+        // Avoid issues with proxies if we populated the object
+        if (isset($data['id']) && !isset($context['object_to_populate'])) {
+            $context['object_to_populate'] = $this->iriConverter->getItemFromIri($data['id'], true);
+        }
+
+        return parent::denormalize($data, $class, $format, $context); // TODO: Change the autogenerated stub
```
Comment (theofidry) [inline src/Serializer/ItemNormalizer.php]: TODO to be removed

File: src/Serializer/ItemNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        $rawData = parent::normalize($object, $format, $context);
+        if (!is_array($rawData)) {
+            return $rawData;
+        }
+
+        if (!isset($data['id'])) {
+            $data['id'] = $this->iriConverter->getIriFromItem($object);
+        }
+
+        return array_merge($data, $rawData);
```
Comment (theofidry) [inline src/Serializer/ItemNormalizer.php]: no better way than an `array_merge`? Not sure about this but isn't iterating over the second array to add values to the first one faster?

File: src/Serializer/ItemNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+class ItemNormalizer extends AbstractItemNormalizer
+{
+    public function normalize($object, $format = null, array $context = [])
+    {
+        $rawData = parent::normalize($object, $format, $context);
+        if (!is_array($rawData)) {
+            return $rawData;
+        }
+
+        if (!isset($data['id'])) {
```
Comment (theofidry) [inline src/Serializer/ItemNormalizer.php]: is ID an IRI in all cases?

File: src/Serializer/ItemNormalizer.php (hunk 1 above)
Comment (dunglas) [inline src/Serializer/ItemNormalizer.php]: Good catch

File: src/Serializer/ItemNormalizer.php (hunk 2 above)
Comment (dunglas) [inline src/Serializer/ItemNormalizer.php]: I want the ID to be first.

File: src/Serializer/ItemNormalizer.php (hunk 3 above)
Comment (dunglas) [inline src/Serializer/ItemNormalizer.php]: IMO it's better to always use an IRI as identifier but I'm open to discussion :)

File: src/Serializer/ItemNormalizer.php (hunk 2 above)
Comment (theofidry) [inline src/Serializer/ItemNormalizer.php]: how does the array_merge guarantees that?

File: src/Serializer/ItemNormalizer.php (hunk 3 above)
Comment (theofidry) [inline src/Serializer/ItemNormalizer.php]: it's alright, just wanted to make sure

Comment (Simperfit) [inline src/Serializer/ItemNormalizer.php]: I've seen a discussion about this https://github.com/api-platform/core/issues/612

Comment (teohhanhui) [inline src/Serializer/ItemNormalizer.php]: @Simperfit I don't think #612 is related.

File: src/Serializer/ItemNormalizer.php (hunk 2 above)
Comment (teohhanhui) [inline src/Serializer/ItemNormalizer.php]: Where does `$data` come from?

**Pull request merged**